File: app/src/main/java/com/base/launcher/util/FileUtils.java
package com.base.launcher.util;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

public class FileUtils {

    private static final String TAG = "BaseFileUtil";

    public static boolean copyAssetsToSDCard(Context context, String assetDir, String targetPath) {
        Log.d(TAG, "copyAssetsToSDCard: assetDir=" + assetDir + ", targetPath=" + targetPath);
        try {
            File targetFile = new File(targetPath);
            if (!targetFile.exists()) {
                createFile(targetPath);
            }

            AssetManager assetManager = context.getAssets();
            String[] fileNames = assetManager.list(assetDir);
            if (fileNames != null) {
                for (String fileName : fileNames) {
                    InputStream inputStream = assetManager.open(assetDir + "/" + fileName);
                    File outputFile = new File(targetPath, fileName);
                    Log.d(TAG, "outputFile: " + outputFile.getAbsolutePath());

                    if (!outputFile.exists()) {
                        createFile(outputFile.getAbsolutePath());

                        FileOutputStream outputStream = new FileOutputStream(outputFile);
                        try {
                            byte[] buffer = new byte[4096];
                            int length;
                            while ((length = inputStream.read(buffer)) > 0) {
                                outputStream.write(buffer, 0, length);
                            }
                            outputStream.flush();
                        } finally {
                            try {
                                inputStream.close();
                            } catch (IOException ignored) {
                            }
                            try {
                                outputStream.close();
                            } catch (IOException ignored) {
                            }
                        }
                    } else {
                        // If file already exists, still close the inputStream
                        try {
                            inputStream.close();
                        } catch (IOException ignored) {
                        }
                    }
                }
            }
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    public static File createFile(String path) {
        if (path == null || path.trim().isEmpty()) {
            return null;
        }

        File target = new File(path);
        try {
            // Handle directory path (ends with separator)
            if (path.endsWith(File.separator)) {
                Log.d(TAG, "target.mkdirs()....");
                return target.mkdirs() ? target : null;
            }

            // Handle file path
            if (!createParentDirs(target)) {
                return null;
            }
            if (target.isDirectory() || target.createNewFile()) {
                return target;
            }
        } catch (IOException e) {
            return null;
        }
        return null;
    }

    private static boolean createParentDirs(File file) throws IOException {
        File parent = file.getParentFile();
        if (parent != null && !parent.exists()) {
            if (!parent.mkdirs()) {
                return false;
            }
        }
        return true;
    }
}

File: app/src/main/java/com/base/launcher/util/FileUtils.kt
package com.base.launcher.util

import android.content.Context
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object FileUtils {
    private const val TAG = "BaseFileUtil"

    @JvmStatic
    fun copyAssetsToSDCard(context: Context, assetDir: String, targetPath: String): Boolean {
        Log.d(TAG, "copyAssetsToSDCard: assetDir=$assetDir, targetPath=$targetPath")
        return try {
            val targetFile = File(targetPath)
            if (!targetFile.exists()) createFile(targetPath)

            val assetManager = context.assets
            val fileNames = assetManager.list(assetDir)
            if (fileNames != null) {
                for (fileName in fileNames) {
                    val inputStream = assetManager.open("$assetDir/$fileName")
                    val outputFile = File(targetPath, fileName)
                    Log.d(TAG, "outputFile: ${outputFile.absolutePath}")
                    if (!outputFile.exists()) {
                        createFile(outputFile.absolutePath)
                        val outputStream = FileOutputStream(outputFile)
                        try {
                            val buffer = ByteArray(4096)
                            var length: Int
                            while (inputStream.read(buffer).also { length = it } > 0) {
                                outputStream.write(buffer, 0, length)
                            }
                            outputStream.flush()
                        } finally {
                            try { inputStream.close() } catch (_: IOException) {}
                            try { outputStream.close() } catch (_: IOException) {}
                        }
                    } else {
                        try { inputStream.close() } catch (_: IOException) {}
                    }
                }
            }
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    @JvmStatic
    fun createFile(path: String?): File? {
        if (path.isNullOrBlank()) return null
        val target = File(path)
        return try {
            if (path.endsWith(File.separator)) {
                Log.d(TAG, "target.mkdirs()....")
                if (target.mkdirs()) target else null
            } else {
                if (!createParentDirs(target)) return null
                if (target.isDirectory || target.createNewFile()) target else null
            }
        } catch (_: IOException) {
            null
        }
    }

    @Throws(IOException::class)
    private fun createParentDirs(file: File): Boolean {
        val parent = file.parentFile
        return parent == null || parent.exists() || parent.mkdirs()
    }
}
